### PR TITLE
Fix ugly semantic coloring on CSL languages while editing

### PR DESCRIPTION
**Pull request by lkishalmi**
This one is bugging me for years.

Whenever I insert or write text to a CSL based editor the semantic coloring does not keep track with the document change
and some text elements get  discolored as you type. The next parsing run would then set the colors right.

See the discoloration of `permissions` before:
[Screencast From 2024-12-01 13-04-47.webm](https://github.com/user-attachments/assets/6be42a1e-e5ef-4b2e-9810-c1c2eaf923b8)

After this fix:
[Screencast From 2024-12-01 13-07-58.webm](https://github.com/user-attachments/assets/0c94ec55-0d84-4b7b-ac61-6e566126e47f)

Fortunately or Unfortunately , I had to change the code here and there in order to be able to understand what's going on.

So this could be a one -liner, removing the `Math.min` form:
https://github.com/apache/netbeans/blob/c2ec6e4d37b9195017e8df82406bfab6b2d29d95/ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java#L322

I submitted my whole work on this though, hoping that it would simplify that part of CSL and makes it less cryptic.

So my changes in brief (read it with the code changes):
  1. Updated the language level to Java 17
  2. Made `SequenceElement` as a record. The `ComparisonItem` got me a lot of headache, at the end it was only used to get the `tailSet` method work. Interestingly that subset just used as an optimization, the editor works even if the highlight sequence return the whole sequence all the time. (Tested with a 3500+ lines yaml, the performance overhead was barely noticeable.) Yet, I've copied the `SortedSet` into a list, where a simple binary search could be used to get a partial iterator.
  3. Using just the iterator on `GsfHighlightSequence` made it a lot simpler.
  4. Initially, I was thinking that the issue is due to some failed/garbage collected weak listener creation/registration, so I moved a few things to lambdas there.
  5. Removed a lot of commented/unused code inherited from Retouche (I think that was the code name of the Java Editor rewrite)

### CODE REVIEW

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java [hunk 1]
```diff
@@ -233,101 +187,61 @@ public int getShiftedPos(int pos) {
...
         @Override
         public AttributeSet getAttributes() {
-            return (element != null)
-                    ? layer.getColoring(element.coloring, element.language)
-                    : SimpleAttributeSet.EMPTY;
+            return getColoring(element.coloring(), element.language());
         }
     }
+
+    private static <T> int firstSequenceElement(List<SequenceElement> l, int offset) {
```
Comment (matthiasblaesing) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:228]: A comment what this method should do would be nice. I have a feeling, but knowing would be nice.

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java [hunk 2]
```diff
@@ -53,19 +50,17 @@
  * 
  * @author Tor Norbye
  */
-public class GsfSemanticLayer extends AbstractHighlightsContainer implements DocumentListener {
+public final class GsfSemanticLayer extends AbstractHighlightsContainer implements DocumentListener {
     
     // -J-Dorg.netbeans.modules.csl.editor.semantic.GsfSemanticLayer.level=FINE
     private static final Logger LOG = Logger.getLogger(GsfSemanticLayer.class.getName());
```
Comment (matthiasblaesing) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:56]: `LOG` is unused and can be removed (reduce warnings).

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java [hunk 3]
```diff
@@ -53,19 +50,17 @@
...
+    private List<SequenceElement> colorings = List.of();
     private List<Edit> edits;
-    private Map<Language,Map<Coloring, AttributeSet>> CACHE = new HashMap<Language,Map<Coloring, AttributeSet>>();
-    private Document doc;
-    private List<Lookup.Result> coloringResults = new ArrayList<Lookup.Result>(3);
-    private List<LookupListener> coloringListeners = new ArrayList<LookupListener>(3);
+    private final Map<Language,Map<Coloring, AttributeSet>> cache = new HashMap<>();
+    private final Document doc;
+    private final List<Lookup.Result> coloringResults = new ArrayList<>(3);
+    private final List<LookupListener> coloringListeners = new ArrayList<>(3);
```
Comment (matthiasblaesing) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:63]: My take on the reason for this:

```suggestion
    // Write only - used to keep strong reference to Lookup.Result/LookupListener
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private final List<Lookup.Result> coloringResults = new ArrayList<>(3);
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private final List<LookupListener> coloringListeners = new ArrayList<>(3);
```

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/HighlightImpl.java
Comment (matthiasblaesing) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/HighlightImpl.java:1]: Suggested by NB:

- make all fields final
- use `StringBuilder` in `coloringsToString` instead of `StringBuffer`
- modifiy `coloringsAttributesOrder` to `private static final Collection<ColoringAttributes> coloringsAttributesOrder = List.of(`
- in `parse` drop the reassignment of `a` and move the `a.trim()` into the `valueOf` call

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/SemanticHighlighter.java [hunk 4]
```diff
@@ -133,139 +109,39 @@ public final void cancel() {
...
+        if (language != null) {
+            ColoringManager manager = language.getColoringManager();
+            SemanticAnalyzer task = language.getSemanticAnalyzer();
+            if (manager != null && task != null) {
+                Parser.Result r = resultIterator.getParserResult();
+                if (r instanceof ParserResult) {
+                    process(language, (ParserResult) r, newColoring);
+                }
+            }
+        }
```
Comment (matthiasblaesing) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/SemanticHighlighter.java:133]: I think this can be simplified to (`manager` and `task` can't be null, at least that is my reading from the code and assured by api annotations).

```suggestion
        if (language != null) {
            Parser.Result r = resultIterator.getParserResult();
            if (r instanceof ParserResult parserResult) {
                process(language, parserResult, newColoring);
            }
        }
```

**Review by matthiasblaesing — COMMENTED** [review]
Looks generally sane to me (only eyeballed). My comments are mostly stylistic (you opened the can of worms ;-)).

The one place I'd really like to see improved is documentation, what `firstSequenceElement` should be doing.

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java [hunk 5]
```diff
@@ -233,101 +187,61 @@ public int getShiftedPos(int pos) {
...
+            return getShiftedPos(element.range().getStart());
         }
 
         @Override
         public int getEndOffset() {
-            return (element != null)
-                    ? Math.min(layer.getShiftedPos(element.range.getEnd()), nextElementStartOffset)
-                    : Integer.MAX_VALUE;
+            return getShiftedPos(element.range().getEnd());
         }
```
Comment (mbien) [inline ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:220]: wondering if you could move the fix into its own commit? Even if its just the one liner which removes `Math.min`.

It would get lost among the other improvements. We do also have a rule in the reviewer guide which states to try to avoid obfuscating fixes with code refactoring. (I know I did break it myself too a few times, but I believe this case here is easy to split still)

**Review by mbien — APPROVED** [review]
Code looks good. Matthias made some good suggestions though.

I vaguely remember the old "schlieman" API impl having a similar problem. The bug could be potentially quite old. Good job hunting it down!

File: ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java (hunk 5 above)
Comment (lkishalmi) [reply ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:220]: Well, I probably close this PR and create two new ones. One with the one-line fix, the other is with the code cleanups.

While fixing the issues on  `HighlightImp`, it turned out that it's not used at all.
I guess it is just an old implementation copied over from `...java.semantic` package (that one is used), and later it turned out CSL would not use token based Highlighting.

Comment (mbien) [reply ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java:220]: your choice. Commit messages is all what is left in the repo history once merged. If they describe what they do -> all good. But if you want to split this into two PRs -> also good of course.

**Comment by lkishalmi** [comment]
Closing in favor of #8018
